relay-server/src/actors/processor.rs

```diff
@@ -2237,18 +2240,95 @@ impl EnvelopeProcessorService {
+                self.tag_error_with_sampling_decision(state);
+            }
+            EventType::Transaction => {
+                self.compute_sampling_decision(state);
+            }
+            _ => {}
+        }
+    }
+
+    /// Computes the sampling decision on the incoming transaction.
     fn compute_sampling_decision(&self, state: &mut ProcessEnvelopeState) {
-        state.sampling_result = utils::should_keep_event(
+        state.sampling_result = utils::get_sampling_result(
             self.config.processing_enabled(),
-            &state.project_state,
+            Some(&state.project_state),
             state.sampling_project_state.as_deref(),
             state.envelope().dsc(),
             state.event.value(),
             state.envelope().meta().client_addr(),
         );
     }
 
+    /// Runs dynamic sampling on an incoming error and tags it in case of successful sampling
+    /// decision.
+    ///
+    /// This execution of dynamic sampling is technically a "simulation" since we will use the result
+    /// only for tagging errors and not for actually sampling incoming events.
+    fn tag_error_with_sampling_decision(&self, state: &mut ProcessEnvelopeState) {
+        let mut sampling_result = None;
+        // We want to run dynamic sampling only if we have a root project state and a dynamic
+        // sampling context.
+        //
+        // In reality the dynamic sampling logic supports optional root state and dsc but it will
+        // return keep. In our case having a keep in case of none root state and dsc will be
+        // a problem, since in reality we can't infer anything without trace metadata.
+        if let (Some(root_project_state), Some(dsc)) = (
+            state.sampling_project_state.as_deref(),
+            state.envelope().dsc(),
+        ) {
+            sampling_result = Some(utils::get_sampling_result(
+                self.config.processing_enabled(),
+                None,
+                Some(root_project_state),
+                Some(dsc),
+                None,
+                state.envelope().meta().client_addr(),
+            ));
+        }
+
+        // In case the sampling result is positive, we assume that all the transactions
+        // that have this DSC will be sampled and thus we mark the error as "having
+        // a full trace".
+        match state.event.value_mut() {
+            Some(event) => {
+                // In case we have no contexts object, we have to create it.
+                if event.contexts.value().is_empty() {
+                    event.contexts = Annotated::new(Contexts::new());
+                }
+
+                // We want to get the specific trace context, or we want to create it in case
+                // it is not there.
+                let context = event.contexts.value_mut().as_mut().map(|context| {
+                    context
+                        .get_or_insert_with(TraceContext::default_key(), || Trace(Box::default()))
```

Comment:
Good observation, I forgot about those two lines when adding the tri-state. What we want is to not run the mutation in case the sampling result is None.